File: app/src/main/java/io/scal/secureshareui/controller/SiteController.java

package io.scal.secureshareui.controller;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Message;
import android.webkit.MimeTypeMap;

import net.opendasharchive.openarchive.R;
import net.opendasharchive.openarchive.db.Media;
import net.opendasharchive.openarchive.db.Space;
import net.opendasharchive.openarchive.services.dropbox.DropboxSiteController;
import net.opendasharchive.openarchive.services.webdav.WebDAVSiteController;

import java.io.File;
import java.io.IOException;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;

import static net.opendasharchive.openarchive.services.dropbox.DropboxSiteController.SITE_KEY;

public abstract class SiteController {
    private OnEventListener mPublishEventListener;
    protected Context mContext;
    protected SiteControllerListener mListener;
    protected String mJobId; // this is whatever the app wants it to be, we'll pass it back with our callbacks
    public static final int CONTROLLER_REQUEST_CODE = 101;
    public static final int MESSAGE_TYPE_SUCCESS = 23423430;
    public static final int MESSAGE_TYPE_FAILURE = 23423431;
    public static final int MESSAGE_TYPE_PROGRESS = 23423432;
    public static final String MESSAGE_KEY_TYPE = "message_type";
    public static final String MESSAGE_KEY_JOB_ID = "job_id";
    public static final String MESSAGE_KEY_CODE = "code";
    public static final String MESSAGE_KEY_MESSAGE = "message";
    public static final String MESSAGE_KEY_RESULT = "result";
    public static final String MESSAGE_KEY_PROGRESS = "progress";
    public static final String MESSAGE_KEY_EXCEPTION = "exception";
    public static final String EXTRAS_KEY_DATA = "data";
    public static final String EXTRAS_KEY_USERNAME = "username";
    public static final String EXTRAS_KEY_CREDENTIALS = "credentials";
    public static final String MESSAGE_KEY_STATUS = "status";
    public static final String MESSAGE_KEY_MEDIA_ID = "mediaId";

    public static final String VALUE_KEY_TITLE = "title";
    public static final String VALUE_KEY_SLUG = "slug";
	public static final String VALUE_KEY_BODY = "body";
	public static final String VALUE_KEY_TAGS = "tags";
	public static final String VALUE_KEY_AUTHOR = "author";
	public static final String VALUE_KEY_PROFILE_URL = "profileUrl";
	public static final String VALUE_KEY_LOCATION_NAME = "locationName";
	public static final String VALUE_KEY_MEDIA_PATH = "mediaPath";
	public static final String VALUE_KEY_LICENSE_URL = "licenseUrl";
    public static final String VALUE_KEY_MIME_TYPE = "mimeType";

    
    public static int METADATA_REQUEST_CODE = 24153;

    private static final String TAG = "SiteController";

    public interface OnEventListener {
        public void onSuccess(Space space);

        public void onFailure(Space space, String failureMessage);
        
        public void onRemove(Space space);
    }

    public SiteController(Context context, SiteControllerListener listener, String jobId) {
        mContext = context;
        mListener = listener;
        mJobId = jobId;
    }

    public abstract void startRegistration (Space space);
    public abstract void startAuthentication(Space space);
    
    /**
     * Gives a SiteController a chance to add metadata to the intent resulting from the ChooseAccounts process
     * that gets passed to each SiteController during publishing
     * @param intent   
     */
    public abstract void startMetadataActivity(Intent intent);

    public abstract boolean upload(Space space, Media media, HashMap<String, String> valueMap) throws IOException;
    public abstract boolean delete(Space space, String bucketName, String mediaFile);

    public abstract ArrayList<File> getFolders (Space space, String path)  throws IOException;

    public void cancel() {}

    public static SiteController getSiteController(String site, Context context, SiteControllerListener listener, String jobId) {
       if (site.equals(ArchiveSiteController.SITE_KEY)) {
            return new ArchiveSiteController(context, listener, jobId);
        }
        else if (site.equals(WebDAVSiteController.SITE_KEY))
       {
           try {
               return new WebDAVSiteController(context,listener,jobId);
           } catch (Exception e) {
               e.printStackTrace();
               return null;
           }
       }
       else if (site.equals(SITE_KEY))
       {
           try {
               return new DropboxSiteController(context,listener,jobId);
           } catch (Exception e) {
               e.printStackTrace();
               return null;
           }
       }
       /**
        else if (site.equalsIgnoreCase(PirateBoxSiteController.SITE_KEY)) {
           return new PirateBoxSiteController(context,listener,jobId);
       }**/
        return null;
    }

    public static boolean isAudioFile(File mediaFile) {
    	if(getMimeType(mediaFile).contains("audio")) {
    		return true;
    	}
    	return false;
    }

    public static boolean isImageFile(File mediaFile) {
    	if(getMimeType(mediaFile).contains("image")) {
    		return true;
    	}
    	return false;
    }

    public static boolean isVideoFile(File mediaFile) {
    	if(getMimeType(mediaFile).contains("video")) {
    		return true;
    	}
    	return false;
    }

    private static String getMimeType(File mediaFile) {
    	Uri fileUri = Uri.fromFile(mediaFile);
    	String fileExtension = MimeTypeMap.getFileExtensionFromUrl(fileUri.toString());
    	return MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtension);
    }

    public OnEventListener getOnPublishEventListener() {
        return this.mPublishEventListener;
    }

    public void setOnEventListener(OnEventListener publishEventListener) {
        this.mPublishEventListener = publishEventListener;
    }

    /**
     * result is a site specific unique id that we can use to fetch the data,
     * build an embed tag, etc. for some sites this might be a URL
     *
     * @param result
     */
    public void jobSucceeded(String result) {
        Message msg = new Message();
        Bundle data = new Bundle();
        data.putInt(MESSAGE_KEY_TYPE, MESSAGE_TYPE_SUCCESS);
        data.putString(MESSAGE_KEY_JOB_ID, mJobId);
        data.putString(MESSAGE_KEY_RESULT, result);
        msg.setData(data);
        mListener.success(msg);
    }

    public void jobFailed(Exception exception, int errorCode, String errorMessage) {
        Message msg = new Message();
        Bundle data = new Bundle();
        data.putInt(MESSAGE_KEY_TYPE, MESSAGE_TYPE_FAILURE);
        data.putString(MESSAGE_KEY_JOB_ID, mJobId);
        data.putInt(MESSAGE_KEY_CODE, errorCode);
        data.putString(MESSAGE_KEY_MESSAGE, errorMessage);
        data.putSerializable("exception", (Serializable) exception);
        msg.setData(data);
        mListener.failure(msg);
    }

    public void jobProgress(long contentLengthUploaded, String message) {
        Message msg = new Message();
        Bundle data = new Bundle();
        data.putInt(MESSAGE_KEY_TYPE, MESSAGE_TYPE_PROGRESS);
        data.putString(MESSAGE_KEY_JOB_ID, mJobId);
        data.putLong(MESSAGE_KEY_PROGRESS, contentLengthUploaded);
        data.putString(MESSAGE_KEY_MESSAGE, message);
        msg.setData(data);
        mListener.progress(msg);
    }

    public static int getAccountIcon(String site, boolean isConnected, boolean areCredentialsValid) {
        if (site.equals(ArchiveSiteController.SITE_KEY)) {
            if (!isConnected) {
                return R.drawable.library;
            }
            return areCredentialsValid ? R.drawable.library : R.drawable.library;
        }

        return R.drawable.ic_launcher;
    }

}

File: app/src/main/java/io/scal/secureshareui/controller/SiteController.kt
package io.scal.secureshareui.controller

import android.content.Context
import android.net.Uri
import kotlin.Throws
import org.witness.proofmode.ProofMode
import android.os.Bundle
import android.os.Message
import net.opendasharchive.openarchive.db.Media
import net.opendasharchive.openarchive.db.Space
import net.opendasharchive.openarchive.services.webdav.WebDavSiteController
import net.opendasharchive.openarchive.services.dropbox.DropboxSiteController
import net.opendasharchive.openarchive.util.Prefs
import org.witness.proofmode.crypto.HashUtils
import timber.log.Timber
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.lang.Exception
import java.util.ArrayList
import kotlin.collections.HashMap

abstract class SiteController(
    protected var mContext: Context,
    private var mListener: SiteControllerListener?, // this is whatever the app wants it to be, we'll pass it back with our callbacks
    private var mJobId: String?
) {
    private var mPublishEventListener: OnEventListener? = null

    ///changes
    interface OnEventListener {
        fun onSuccess(space: Space?)
        fun onFailure(space: Space?, failureMessage: String?)
        fun onRemove(space: Space?)
    }

    abstract fun startAuthentication(space: Space?)

    /**
     * Gives a SiteController a chance to add metadata to the intent resulting from the ChooseAccounts process
     * that gets passed to each SiteController during publishing
     */
    @Throws(IOException::class)
    abstract fun upload(space: Space?, media: Media?, valueMap: HashMap<String, String?>): Boolean
    abstract fun delete(space: Space?, bucketName: String?, mediaFile: String?): Boolean

    @Throws(IOException::class)
    abstract fun getFolders(space: Space?, path: String?): ArrayList<File>
    open fun cancel() {}

    fun getProof(media: Media): Array<out File> {
        if (!Prefs.getUseProofMode()) return emptyArray()

        // Don't use geolocation and network information.
        Prefs.putBoolean(ProofMode.PREF_OPTION_LOCATION, false)
        Prefs.putBoolean(ProofMode.PREF_OPTION_NETWORK, false)

        try {
            var hash = ProofMode.generateProof(
                mContext,
                Uri.parse(media.originalFilePath),
                media.mediaHashString
            )

            if (hash == null) {
                val proofHash = HashUtils.getSHA256FromFileContent(
                    mContext.contentResolver.openInputStream(Uri.parse(media.originalFilePath))
                )

                hash = ProofMode.generateProof(mContext, Uri.parse(media.originalFilePath), proofHash)
            }

            return ProofMode.getProofDir(mContext, hash).listFiles() ?: emptyArray()
        }
        catch (exception: FileNotFoundException) {
            Timber.e(exception)

            return emptyArray()
        }
        catch (exception: SecurityException) {
            Timber.e(exception)

            return emptyArray()
        }
    }

    fun setOnEventListener(publishEventListener: OnEventListener?) {
        mPublishEventListener = publishEventListener
    }

    /**
     * result is a site specific unique id that we can use to fetch the data,
     * build an embed tag, etc. for some sites this might be a URL
     *
     */
    fun jobSucceeded(result: String?) {
        val msg = Message()
        val data = Bundle()
        data.putInt(MESSAGE_KEY_TYPE, MESSAGE_TYPE_SUCCESS)
        data.putString(MESSAGE_KEY_JOB_ID, mJobId)
        data.putString(MESSAGE_KEY_RESULT, result)
        msg.data = data
        mListener?.success(msg)
    }

    fun jobFailed(exception: Exception?, errorCode: Int, errorMessage: String?) {
        val msg = Message()
        val data = Bundle()
        data.putInt(MESSAGE_KEY_TYPE, MESSAGE_TYPE_FAILURE)
        data.putString(MESSAGE_KEY_JOB_ID, mJobId)
        data.putInt(MESSAGE_KEY_CODE, errorCode)
        data.putString(MESSAGE_KEY_MESSAGE, errorMessage)
        data.putSerializable("exception", exception)
        msg.data = data
        mListener?.failure(msg)
    }

    fun jobProgress(contentLengthUploaded: Long, message: String?) {
        val msg = Message()
        val data = Bundle()
        data.putInt(MESSAGE_KEY_TYPE, MESSAGE_TYPE_PROGRESS)
        data.putString(MESSAGE_KEY_JOB_ID, mJobId)
        data.putLong(MESSAGE_KEY_PROGRESS, contentLengthUploaded)
        data.putString(MESSAGE_KEY_MESSAGE, message)
        msg.data = data
        mListener?.progress(msg)
    }

    companion object {
        const val CONTROLLER_REQUEST_CODE = 101
        const val MESSAGE_TYPE_SUCCESS = 23423430
        const val MESSAGE_TYPE_FAILURE = 23423431
        const val MESSAGE_TYPE_PROGRESS = 23423432
        const val MESSAGE_KEY_TYPE = "message_type"
        const val MESSAGE_KEY_JOB_ID = "job_id"
        const val MESSAGE_KEY_CODE = "code"
        const val MESSAGE_KEY_MESSAGE = "message"
        const val MESSAGE_KEY_RESULT = "result"
        const val MESSAGE_KEY_PROGRESS = "progress"
        const val EXTRAS_KEY_USERNAME = "username"
        const val EXTRAS_KEY_CREDENTIALS = "credentials"
        const val MESSAGE_KEY_STATUS = "status"
        const val MESSAGE_KEY_MEDIA_ID = "mediaId"
        const val VALUE_KEY_TITLE = "title"
        const val VALUE_KEY_SLUG = "slug"
        const val VALUE_KEY_BODY = "body"
        const val VALUE_KEY_TAGS = "tags"
        const val VALUE_KEY_AUTHOR = "author"
        const val VALUE_KEY_PROFILE_URL = "profileUrl"
        const val VALUE_KEY_LOCATION_NAME = "locationName"
        const val VALUE_KEY_MEDIA_PATH = "mediaPath"
        const val VALUE_KEY_LICENSE_URL = "licenseUrl"
        const val VALUE_KEY_MIME_TYPE = "mimeType"
        var METADATA_REQUEST_CODE = 24153

        @JvmStatic
        fun getSiteController(
            site: String?,
            context: Context?,
            listener: SiteControllerListener?,
            jobId: String?
        ): SiteController? {
            when (site) {
                ArchiveSiteController.SITE_KEY -> return ArchiveSiteController(
                    context!!,
                    listener,
                    jobId
                )
                WebDavSiteController.SITE_KEY -> {
                    return try {
                        WebDavSiteController(context!!, listener, jobId)
                    } catch (e: Exception) {
                        e.printStackTrace()
                        null
                    }
                }
                DropboxSiteController.SITE_KEY -> {
                    return try {
                        DropboxSiteController(context!!, listener, jobId)
                    } catch (e: Exception) {
                        e.printStackTrace()
                        null
                    }
                }
            }
            /**
             * else if (site.equalsIgnoreCase(PirateBoxSiteController.SITE_KEY)) {
             * return new PirateBoxSiteController(context,listener,jobId);
             * } */
            return null
        }
    }
}